service/src/test/java/bio/terra/workspace/common/utils/MdcHookTest.java

```diff
@@ -14,147 +19,228 @@ class MdcHookTest extends BaseUnitTest {
-
-    private void assertMatchesExpectedContext() {
-      try {
-        assertEquals(expectedContext, MDC.getCopyOfContextMap());
-      } catch (AssertionError error) {
-        // Rethrow an AssertionError as an Exception so that Stairway can handle it
-        throw new RuntimeException(error);
-      }
-    }
   }
 
   /** A {@link Step} that always fails in a non-retryable way. */
-  public static class ErrorStep implements Step {
+  public static class ErrorStep extends MdcAssertionStep {
+
+    public static final Exception EXPECTED_EXCEPTION =
+        new RuntimeException("Expected exception on ErrorStep.doStep");
+
+    public ErrorStep(
+        Map<String, String> expectedFlightContext,
+        Integer expectedStepNumber,
+        Direction expectedUndoDirection) {
+      super(expectedFlightContext, expectedStepNumber, expectedUndoDirection);
+    }
+
     @Override
     public StepResult doStep(FlightContext flightContext) {
-      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL);
+      assertMatchesExpectedContext(Direction.DO);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, EXPECTED_EXCEPTION);
     }
 
     @Override
     public StepResult undoStep(FlightContext flightContext) {
+      assertMatchesExpectedContext(expectedUndoDirection);
       return StepResult.getStepResultSuccess();
     }
   }
+
+  private abstract static class MdcAssertionStep implements Step {
+    private final Map<String, String> expectedContextBase;
+    final Direction expectedUndoDirection;
+
+    MdcAssertionStep(
+        Map<String, String> expectedFlightContext,
+        Integer expectedStepNumber,
+        Direction expectedUndoDirection) {
+      this.expectedContextBase =
+          ImmutableMap.<String, String>builder()
+              .putAll(expectedFlightContext)
+              .put(MdcHook.FLIGHT_STEP_CLASS_KEY, getClass().getName())
+              .put(MdcHook.FLIGHT_STEP_NUMBER_KEY, expectedStepNumber.toString())
+              .build();
+      this.expectedUndoDirection = expectedUndoDirection;
+    }
+
+    void assertMatchesExpectedContext(Direction flightStepDirection) {
+      try {
+        var expectedContext =
+            ImmutableMap.<String, String>builder()
+                .putAll(expectedContextBase)
+                .put(MdcHook.FLIGHT_STEP_DIRECTION_KEY, flightStepDirection.name())
+                .build();
+        assertEquals(expectedContext, MDC.getCopyOfContextMap());
+      } catch (AssertionError error) {
+        // Rethrow an AssertionError as an Exception so that Stairway can handle it
+        throw new RuntimeException(error);
+      }
+    }
+  }
+
+  /**
+   * Submits the flight and polls every 100 ms for its completion, up to 3 seconds.
+   *
+   * @param flightClass to submit with flightId and flightMap instance variables
+   * @return FlightState of the completed flight
+   */
+  private FlightState submitAndWait(Class<? extends Flight> flightClass) throws Exception {
```

Comment:
Note -- before, we would sleep for 3 seconds after every test flight, which is a code smell in tests and made the overall runtime of this unit test class longer than it needed to be.  By polling frequently for flight completion up to 3 seconds, I was seeing each test run in under a second.

Reply:
```
       ,
     0/
    /7,
  .'(
'^ / >
  / <  
  `
```